### PR TITLE
Clearing the cache on re-join

### PR DESCRIPTION
**Pull request by p-schuler**
# PR for issue #1024

## What is being addressed

When receiving a re-join of an already cached device in a multi gw scenario, the gateways that lose the join processing race, will be left with a stale session key, making it impossible to process messages until the cache is refreshed.

## How is this addressed

We clear that particular device from the cache, when we lose the race during the join. That will make sure the session key will be re-fetched on the next message. Note: there is a chance that the device won't get the request at all and then the session keys will only be refreshed on the periodic cache refresh we perform.

Closing #1024

### CODE REVIEW

File: Tests/Unit/NetworkServer/LoRaDeviceRegistryTest.cs [hunk 1]
```diff
@@ -44,6 +44,28 @@ public async Task GetDeviceForJoinRequestAsync_When_Device_Api_Throws_Error_Shou
...
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task GetDeviceForJoinRequestAsync_When_Join_Handled_By_Other_Cache_Is_Updated(bool joinedDevice)
+        {
+            var devNonce = new DevNonce(1);
+            var apiService = new Mock<LoRaDeviceAPIServiceBase>();
+            var otaaDevice = TestDeviceInfo.CreateOTAADevice(1);
+            if (joinedDevice) otaaDevice.AppSKey = "123";
```
Comment (bastbu) [inline Tests/Unit/NetworkServer/LoRaDeviceRegistryTest.cs:55]: [nit] you could create the key here using `TestKeys` (will make the build green again)

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceRegistry.cs [hunk 2]
```diff
@@ -215,6 +215,15 @@ public async Task<LoRaDevice> GetDeviceForJoinRequestAsync(string devEUI, DevNon
 
             if (searchDeviceResult.IsDevNonceAlreadyUsed)
             {
+                // another gateway processed the join request. If we have it in the cache
+                // with existing session keys, we need to invalidate that entry, to ensure
+                // it gets re-fetched on the next message
+                if (this.deviceCache.TryGetByDevEui(devEUI, out var someDevice) && someDevice.AppSKey != null)
```
Comment (bastbu) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceRegistry.cs:221]: I don't understand the case `someDevice.AppSKey != null`, why would we have it in the cache without an `AppSKey`?

**Review by bastbu — APPROVED** [review]

File: Tests/Unit/NetworkServer/LoRaDeviceRegistryTest.cs (hunk 1 above)
Comment (p-schuler) [reply Tests/Unit/NetworkServer/LoRaDeviceRegistryTest.cs]: Already fixed. Thanks. Didn't know about the TestKeys. Will use them when I need real keys.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceRegistry.cs (hunk 2 above)
Comment (p-schuler) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceRegistry.cs:221]: As discussed: when a join request is received, we do put it into the cache without the session keys yet initialized. That means, we can't handle incoming messages, but we can use it for other join messages (if we miss the join receive window).

**Pull request merged**